lib/oli_web/live/delivery/student_dashboard/course_content_live.ex

```diff
@@ -0,0 +1,349 @@
+        [
+          {socket.assigns.current_level + 1, selected_resource_index,
+           get_resource_prefix(
+             current_node,
+             socket.assigns.section.display_curriculum_item_numbering
+           )}
+        ]
+
+    socket =
+      socket
+      |> update(:current_level, &(&1 + 1))
+      |> assign(:current_position, selected_resource_index)
+      |> assign(:current_level_nodes, current_node["children"])
+      |> assign(:breadcrumbs_tree, breadcrumbs_tree)
+
+    {:noreply, socket}
+  end
+
+  @impl Phoenix.LiveView
+  def handle_event(
+        "breadcrumb-navigate",
+        %{"target_level" => target_level, "target_position" => target_position},
+        socket
+      ) do
+    breadcrumbs_tree =
+      update_breadcrumbs_tree(
+        socket.assigns.breadcrumbs_tree,
+        String.to_integer(target_level),
+        String.to_integer(target_position)
+      )
+
+    current_level_nodes = get_current_level_nodes(breadcrumbs_tree, socket.assigns.hierarchy)
+
+    {level, current_position, _text} = List.last(breadcrumbs_tree)
+
+    socket =
+      socket
+      |> assign(:breadcrumbs_tree, breadcrumbs_tree)
+      |> assign(:current_level, level + 1)
+      |> assign(:current_position, current_position)
+      |> assign(:current_level_nodes, current_level_nodes)
+
+    {:noreply, socket}
+  end
+
+  @impl Phoenix.LiveView
+  def handle_event(
+        "open_resource",
+        %{"resource_slug" => resource_slug, "resource_type" => resource_type},
+        socket
+      ) do
+    {:noreply,
+     redirect(socket,
+       to:
+         Routes.page_delivery_path(
+           socket,
+           String.to_existing_atom(resource_type),
+           socket.assigns.section.slug,
+           resource_slug
+         )
+     )}
+  end
+
+  defp update_breadcrumbs_tree(breadcrumbs_tree, 0, _target_position),
+    do: [hd(breadcrumbs_tree)]
+
+  defp update_breadcrumbs_tree(breadcrumbs_tree, target_level, target_position) do
+    {true, breadcrumbs_tree} =
+      Enum.reduce(breadcrumbs_tree, {false, []}, fn b, acc ->
+        {level, position, _text} = b
+        {found, breadcrumbs_tree} = acc
+
+        if level == target_level and position == target_position and found == false do
+          {true, breadcrumbs_tree}
+        else
+          if found == true do
+            {true, breadcrumbs_tree}
+          else
+            {false, [b | breadcrumbs_tree]}
+          end
+        end
+      end)
+
+    Enum.reverse(breadcrumbs_tree)
+  end
+
+  defp get_current_level_nodes(breadcrumbs_tree, hierarchy) do
+    {0, current_level_nodes} =
+      breadcrumbs_tree
+      |> Enum.reduce({length(breadcrumbs_tree) - 1, hierarchy["children"]}, fn b, acc ->
+        {steps_remaining, hierarchy} = acc
+        {_level, position, _text} = b
+
+        if steps_remaining == 0 do
+          {0, hierarchy}
+        else
+          {steps_remaining - 1, Enum.fetch!(hierarchy, position)["children"]}
+        end
```

Comment:
Just a minor writing detail, maybe this could be written in one line, with something like: `if steps_remaining == 0, do: {0, hierarchy}, else: {steps_remaining - 1, Enum.fetch!(hierarchy, position)["children"]}`